### PR TITLE
Fused attention fixes for cuDNN 8.9.3

### PR DESCRIPTION
**Pull request by cyanguwa**
This PR fixes issues with the cudnn frontend code after upgrading cuDNN to 8.9.3. Mostly, there were two issues, 1) incorrect results with bprop for >=512 seqlen, and 2) runtime errors with the bias graph for <=512 seqlen.

### CODE REVIEW

**Comment by zlsh80826** [comment]
Hello Charlene,
With those fixes, the flash attention will not be compatible with cudnn < 8.9.3. Also, the flash attention kernels before 8.9.3 seem not feasible. Should we change the minimum cudnn requirements of flash attention?

**Comment by cyanguwa** [comment]
> Hello Charlene, With those fixes, the flash attention will not be compatible with cudnn < 8.9.3. Also, the flash attention kernels before 8.9.3 seem not feasible. Should we change the minimum cudnn requirements of flash attention?

Yes, I will do that. Thanks!

**Comment by cyanguwa** [comment]
Pipeline 8937934

**Comment by cyanguwa** [comment]
Pipeline 8937934 has passed.

**Comment by zlsh80826** [comment]
Hi @cyanguwa,
Fused attention (max 512) is good after 8.9.1. The wrong dbias was introduced in 8.9.3 internal version and will not be released to public, so fused attention is good across 8.9.1 to 8.9.3.

**Comment by cyanguwa** [comment]
Pipeline 8967304 has passed.

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu [hunk 1]
```diff
@@ -1185,7 +1199,7 @@ void fused_attn_arbitrary_seqlen_fwd_qkvpacked(
 
     // QKV shape is [b, s, 3, h, d]
     void *devPtrQKV = input_QKV->data.dptr;
-    const auto stride = num_head * head_dim;
+    const auto stride = 2 * num_head * head_dim;
```
Comment (timmoon10) [inline transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu:1202]: Just to clarify, we are assuming that the input data is fp16/bf16?

**Review by timmoon10 — APPROVED** [review]

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu (hunk 1 above)
Comment (cyanguwa) [reply transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu:1202]: Yes, these two versions of fused_attn (m512 and arbitrary) would only run with FP16/BF16 data.

**Comment by cyanguwa** [comment]
Just to clarify, pipeline 8967304 included all the changes in this PR. I had to make some LD_LIBRARY_PATH changes because the cuDNN version used in our current PyTorch CI container breaks the code, but it's only an internal build, and the more recent build (the one I used) would pass. Hence the add debug info ([2ac2d48](https://github.com/NVIDIA/TransformerEngine/pull/311/commits/2ac2d486f79dbb89c59f138c0b2b563d3d8b05c4)) and then later delete ([31bd096](https://github.com/NVIDIA/TransformerEngine/pull/311/commits/31bd096317d23e931b853aa54db80a1382961ed9)).

I also tested the earlier versions (publicly released versions, not internal builds) of cuDNN with `test_fused_attention.py`, and the requirements for m512 fused attention (8.9.1+) and arbitrary flash attention (8.9.0+) seem to stand. Hence the macro changes regarding the `CUDNN_VERSION`s.

**Pull request merged**